webhooks/virtualmachine/validation/virtualmachine_validator.go

```diff
@@ -1951,3 +1953,169 @@ func (v validator) validateGroupName(
+								p.Child("operator"),
+								expr.Operator,
+								[]metav1.LabelSelectorOperator{metav1.LabelSelectorOpIn}))
+						}
+					}
+				}
+
+				if rs.TopologyKey != topology.KubernetesTopologyZoneLabelKey {
+					allErrs = append(allErrs, field.NotSupported(
+						p.Child("topologyKey"), rs.TopologyKey, []string{topology.KubernetesTopologyZoneLabelKey}))
+				}
+			}
+		}
+	}
+
+	if a := affinity.VMAntiAffinity; a != nil {
+		p := path.Child("vmAntiAffinity")
+
+		if len(a.RequiredDuringSchedulingIgnoredDuringExecution) > 0 {
+			allErrs = append(allErrs, field.Forbidden(p.Child("requiredDuringSchedulingIgnoredDuringExecution"),
+				"VM anti-affinity with RequiredDuringSchedulingIgnoredDuringExecution is not allowed"))
+		}
+
+		if len(a.PreferredDuringSchedulingIgnoredDuringExecution) > 0 {
+			p := p.Child("preferredDuringSchedulingIgnoredDuringExecution")
+
+			for idx, rs := range a.PreferredDuringSchedulingIgnoredDuringExecution {
+				p := p.Index(idx)
+
+				if rs.LabelSelector != nil {
+					p := p.Child("labelSelector")
+
+					for exprIdx, expr := range rs.LabelSelector.MatchExpressions {
+						p := p.Child("matchExpressions").Index(exprIdx)
+
+						if expr.Operator != metav1.LabelSelectorOpIn {
+							allErrs = append(allErrs, field.NotSupported(
+								p.Child("operator"),
+								expr.Operator,
+								[]metav1.LabelSelectorOperator{metav1.LabelSelectorOpIn}))
+						}
+					}
+				}
+
+				if rs.TopologyKey != topology.KubernetesTopologyZoneLabelKey {
+					allErrs = append(allErrs, field.NotSupported(
+						p.Child("topologyKey"), rs.TopologyKey, []string{topology.KubernetesTopologyZoneLabelKey}))
+				}
+			}
+		}
+
+		if len(a.RequiredDuringSchedulingPreferredDuringExecution) > 0 {
+			allErrs = append(allErrs, field.Forbidden(p.Child("requiredDuringSchedulingPreferredDuringExecution"),
+				"VM anti-affinity with RequiredDuringSchedulingPreferredDuringExecution is not allowed"))
+		}
+
+		if len(a.PreferredDuringSchedulingPreferredDuringExecution) > 0 {
+			p := p.Child("preferredDuringSchedulingPreferredDuringExecution")
+
+			for idx, rs := range a.PreferredDuringSchedulingPreferredDuringExecution {
+				p := p.Index(idx)
+
+				if rs.LabelSelector != nil {
+					if len(rs.LabelSelector.MatchExpressions) != 0 {
+						allErrs = append(allErrs, field.Forbidden(
+							p.Child("labelSelector", "matchExpressions"), "MatchExpressions is not allowed"))
+					}
+				}
+
+				if rs.TopologyKey != "" && rs.TopologyKey != topology.KubernetesTopologyHostLabelKey {
+					allErrs = append(allErrs, field.NotSupported(
+						p.Child("topologyKey"), rs.TopologyKey, []string{"", topology.KubernetesTopologyHostLabelKey}))
+				}
+			}
+		}
+	}
+
+	return allErrs
+}
+
+func (v validator) validateImmutableVMAffinity(
+	_ *pkgctx.WebhookRequestContext,
+	vm, oldVM *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+
+	if !equality.Semantic.DeepEqual(vm.Spec.Affinity, oldVM.Spec.Affinity) {
```

Comment:
IDK how far do we want to take this, but there can potentially be a bunch of lists in the affinity rules (MatchExpressions) and changing their order should not reject the request ideally :(